File: Microverse Challenges/module 3/3.3 hash-tables-Trast00/challenge.js
function hashTable(arr) {
  // write your code here
  const hashmap = new Map()
  
  arr.forEach(element => {
    const key = Math.abs(element%11)
    let newValue = []
    if(hashmap.has(key)){
      newValue = hashmap.get(key)
    }
    newValue.push(element)
    hashmap.set(key, newValue)
  });

  let result = []
  for(let i=0; i<11; i++){
    if(hashmap.has(i)){
      result = [...result, ...hashmap.get(i)]
    }
  }
  
  return result
}

console.log(hashTable([12, 24, 125, 5, 91, 1106, 2, 1021, 29, 3536, 10]))
// => [12, 24, 2, 91, 125, 5, 3536, 1106, 29, 1021, 10]

console.log(hashTable([2, 341, 73, 8265, 234004, 602, 7400000, 200000000]))
// => [341, 234004, 2, 200000000, 7400000, 8265, 73, 602]

module.exports = hashTable

File: Microverse Challenges/module 4/4.3 breadth-first-search-Trast00/challenge.rb
/* Time: O(V + E) Space: O(V) */
def bfs(graph)
  # write your code here
  
  # create a array taht contain the first key
  @result = [0]
  @queue = [0]
  @discovered = [0]


  until @queue.empty? 
    key = @queue.shift
    @discovered << key

    @result = [*@result, *graph[key]].uniq
    graph[key].uniq.each do |element|
      @queue << element if !@discovered.include?(element)
    end
  end

  @result
end

p bfs({
  0 => [2], 
  1 => [4], 
  2 => [5, 0, 3], 
  3 => [2], 
  4 => [1, 5], 
  5 => [4, 2]
})
# => [0, 2, 5, 3, 4, 1]
